src/functions.jl

```diff
@@ -1,407 +1,1444 @@
+ 0
+```
+
+## Specific calls
+
+When giving a _Time for `t`, the function takes a vector as input for `x` and `p`.
+
+- state dimension: 1
+- time dependence: autonomous
+
+```@example
+julia> Hv = HamiltonianVectorField((x, p) -> [x + p, x - p], state_dimension=1)
+julia> Hv(_Time(0), 1, 1)
+ERROR: MethodError: no method matching (::HamiltonianVectorField{:autonomous, 1})(::CTBase._Time, ::Int64, ::Int64)
+julia> Hv(_Time(0), [1], [1])
+2-element Vector{Int64}:
+ 2
+ 0
+```
+
+- state dimension: 1
+- time dependence: nonautonomous
+
+```@example
+julia> Hv = HamiltonianVectorField((t, x, p) -> [t + x + p, x - p], state_dimension=1, time_dependence=:nonautonomous)
+julia> Hv(_Time(1), 1, 1)
+ERROR: MethodError: no method matching (::HamiltonianVectorField{:nonautonomous, 1})(::CTBase._Time, ::Int64, ::Int64)
+julia> Hv(_Time(1), [1], [1])
+2-element Vector{Int64}:
+ 3
+ 0
+```
+
+- state dimension: 2
+- time dependence: autonomous
+
+```@example
+julia> Hv = HamiltonianVectorField((x, p) -> [x[1]^2 + p[2]^2, x[2]^2 - p[1]^2], state_dimension=2)
+julia> Hv(_Time(0), [1, 0], [0, 1])
+2-element Vector{Int64}:
+ 2
+ 0
+```
+
+- state dimension: 2
+- time dependence: nonautonomous
+
+```@example
+julia> Hv = HamiltonianVectorField((t, x, p) -> [t + x[1]^2 + p[2]^2, x[2]^2 - p[1]^2], state_dimension=2, time_dependence=:nonautonomous)
+julia> Hv(_Time(1), [1, 0], [0, 1])
+2-element Vector{Int64}:
+ 3
+ 0
+```
+"""
+struct HamiltonianVectorField{time_dependence, state_dimension}
+    f::Function
+    function HamiltonianVectorField(f::Function; 
+        state_dimension::Union{Symbol,Dimension}=__state_dimension(), 
+        time_dependence::Union{Nothing,Symbol}=__fun_time_dependence())
+        if time_dependence ∉ [:autonomous, :nonautonomous]
+            throw(InconsistentArgument("time_dependence must be either :autonomous or :nonautonomous"))
         end
-        return y
+        new{time_dependence, state_dimension}(f)
     end
 end
 
+# classical calls
+function (F::HamiltonianVectorField{:nonautonomous, N})(t::Time, x::Union{ctNumber, State}, 
+        p::Union{ctNumber, Adjoint}, args...; kwargs...)::ctVector where {N}
+    return F.f(t, x, p, args...; kwargs...)
+end
+function (F::HamiltonianVectorField{:autonomous, N})(x::Union{ctNumber, State}, 
+        p::Union{ctNumber, Adjoint}, args...; kwargs...)::ctVector where {N}
+    return F.f(x, p, args...; kwargs...)
+end
+
+# specific calls: inputs are vectors, except times
+# state dimension: 1
+function (F::HamiltonianVectorField{:nonautonomous, 1})(t::_Time, x::State, p::Adjoint, args...; kwargs...)::ctVector
+    return F.f(t.value, x[1], p[1], args...; kwargs...)
+end
+function (F::HamiltonianVectorField{:autonomous, 1})(t::_Time, x::State, p::Adjoint, args...; kwargs...)::ctVector
+    return F.f(x[1], p[1], args...; kwargs...)
+end
+# state dimension: N>1
+function (F::HamiltonianVectorField{:nonautonomous, N})(t::_Time, x::State, p::Adjoint, args...; kwargs...)::ctVector where {N}
```

Comment:
@ocots please note to explain the usage / role of `_Time` during the catch-up (assuming it is associated to the preserved `autonomous` thing)